### PR TITLE
[RFC] System devicetree demo for mps2 an521 board

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
As discussed in architecture working group meetings related to https://github.com/zephyrproject-rtos/zephyr/issues/51830, here is a proof of concept for running a board definition using system devicetree in "hardware model v1" (for information on "v2" and why it will be needed, see https://github.com/zephyrproject-rtos/zephyr/issues/51831).

To test hello world booting on CPU0:

```
cmake -Bbuild -GNinja -DBOARD=sysdt_mps2_an521 -DAPP_DIR=samples/hello_world -DSYSDT_EXECUTION_DOMAIN=/domains/domain_cpu0 share/sysbuild
cmake --build build
ninja -C build/hello_world run
```

We don't have a TF-M target yet to build non-secure firmware, however,
and I haven't tested CPU1 yet. Those are still TODO.

Changelog:

- v4 (2023-02-28): migrated sysbuild functionality into new cmake module, sysdts.cmake
- v3 (2023-02-24): initial sysbuild and SYSDT_OVERLAY_FILE support
- v2 (2023-02-07): completely re-work the build system support to use the concept of execution domains
- v1 (2022-11-15): this didn't use execution domains, and had several extra cmake variables which were removed in v2

### CODE REVIEW

**Review by nordicjm — COMMENTED** [review]
Looks good so far! @tejlmand we will need to discuss how to get this integrated with sysbuild, as of right now if you run this board through it then it fails with an error about `SYSDT_CPU_CLUSTER` not being set.

**Review by microbuilder — COMMENTED** [review]
Looks good to me, just a minor documentation nit.

I'd like to see how we'll express things like MPU partitioning across `mps2_an521` and `mps2_an521_ns` with this approach, where the secure partition always has priority. Alternatively, ownership of something like the I2C/SPI bus. This at least gives us a starting point for expressing those relationships, though.

**Comment by mbolivar-nordic** [comment]
@nordicjm @microbuilder thanks for the reviews. I will update as requested.

For situational awareness, there are a couple of other changes incoming that aren't reflected in the discussion here:

1. Domains support

@galak doesn't like the set up with the extra .dts files for targeting a particular core, and asked me to look into adding support for the 'execution domains' feature of system devicetree instead, and having each board provide some default domains.

If I find this to be workable, the impact to the PR will be:

- new /domains node in the board-level system devicetree, with child nodes for some default domains that configure things like chosen nodes, etc
- sysdtlib support for the same
- build system variable changes, removing ` -DDTS_SOURCE=cpu0.dts -DSYSDT_CPU_CLUSTER=/cpus-cluster@0` in favor of something like `-DSYSDT_DOMAIN=cpu0-app` to use `/domains/cpu0-oapp` as the default domain

2. Fixes and improvements in sysdtlib resulting from specification discussions

I've been discussing and sending various pull requests to the system devicetree specification (issues: https://github.com/devicetree-org/lopper/issues/133, https://github.com/devicetree-org/lopper/issues/130, https://github.com/devicetree-org/lopper/issues/129, https://github.com/devicetree-org/lopper/issues/128, PRs: https://github.com/devicetree-org/lopper/pull/131, https://github.com/devicetree-org/lopper/pull/126), with at least one more big one incoming.

This will change the semantics of the 'simple-bus' handling (I got this wrong, and simple-bus nodes need to be explicitly mapped into each CPU cluster's address-space).

It will also result in more complete sysdtlib support.

I have at least one more large-ish system DT spec PR incoming this week to make the language in there a bit more formal, etc.

I will update this PR after I'm confident we have the spec issues sorted out.

**Review by nordicjm — COMMENTED** [review]
@tejlmand seems this is not compatible with sysbuild: `cmake -GNinja -DBOARD=sysdt_mps2_an521 -DSYSDT_EXECUTION_DOMAIN=/domains/domain_cpu0 -DAPP_DIR=.. -Dhello_world_SYSDT_EXECUTION_DOMAIN=/domains/domain_cpu0 /tmp/aa/zephyr/share/sysbuild/` gives:
```
   *********************************
   * Running CMake for hello_world *
   *********************************

Loading Zephyr default modules (Zephyr repository).
-- Application: /tmp/aa/zephyr/samples/hello_world
-- CMake version: 3.25.2
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.9") found components: Interpreter 
-- Cache files will be written to: /home/jamie/.cache/zephyr
-- Zephyr version: 3.3.0-rc1 (/tmp/aa/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: sysdt_mps2_an521
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.15.0 (/opt/zephyr-sdk-0.15.0)
-- Found toolchain: zephyr 0.15.0 (/opt/zephyr-sdk-0.15.0)
-- Found Dtc: /opt/zephyr-sdk-0.15.0/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
CMake Error at /tmp/aa/zephyr/cmake/modules/dts.cmake:232 (message):
  Found system devicetree file
  /tmp/aa/zephyr/boards/arm/sysdt_mps2_an521/sysdt_mps2_an521.sysdts, but
  SYSDT_EXECUTION_DOMAIN was not defined.  Set this variable to the path to
  the execution domain you wish to use.
Call Stack (most recent call first):
  /tmp/aa/zephyr/cmake/modules/zephyr_default.cmake:113 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /tmp/aa/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:117 (include_boilerplate)
  /home/jamie/sysbuild/zephyr/share/zephyr-package/cmake/zephyr_package_search.cmake:117 (include)
  /home/jamie/sysbuild/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:152 (check_zephyr_package)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!
CMake Error at cmake/modules/sysbuild_extensions.cmake:255 (message):
  CMake configure failed for Zephyr project: hello_world

  Location: /tmp/aa/zephyr/samples/hello_world
Call Stack (most recent call first):
  CMakeLists.txt:59 (ExternalZephyrProject_Add)
```

File: cmake/modules/dts.cmake [hunk 1]
```diff
@@ -69,246 +141,362 @@ foreach(dts_dir ${DTS_ROOT})
...
+# Halt execution early if there is no devicetree or system devicetree.
+#
+
+if(BOARD_HAS_SYSDTS)
+  # When a system devicetree is present, we'll use the output devicetree
+  # in the build directory as DTS_SOURCE.
   #
-  # Run gen_defines.py to create a header file, zephyr.dts, and edt.pickle.
-  #
+  # TODO: how can we support board revisions in system devicetree?
```
Comment (nordicjm) [inline cmake/modules/dts.cmake:255]: If the revision is still provided with the board name, can it not check if `<board>@<revision>.overlay` exists and if so, include that with the main dts file?

File: cmake/modules/dts.cmake [hunk 2]
```diff
@@ -69,246 +141,362 @@ foreach(dts_dir ${DTS_ROOT})
...
+  endif()
+else()
+  # TODO: What to do about non-posix platforms where NOT CONFIG_HAS_DTS (xtensa)?
+  # Drop support for NOT CONFIG_HAS_DTS perhaps?
+
+  set_ifndef(DTS_SOURCE ${BOARD_DIR}/${BOARD}.dts)
+  if(EXISTS ${DTS_SOURCE})
+    # We found a devicetree. Check for a board revision overlay.
+    if(BOARD_REVISION AND EXISTS ${BOARD_DIR}/${BOARD}_${BOARD_REVISION_STRING}.overlay)
+      list(APPEND DTS_SOURCE ${BOARD_DIR}/${BOARD}_${BOARD_REVISION_STRING}.overlay)
```
Comment (nordicjm) [inline cmake/modules/dts.cmake:276]: (From above) similar to this, boards will either be v2 systemdt or v1 non-systemdt so should be able to have v2 specific handling

File: scripts/dts/python-devicetree/src/devicetree/sysdtlib.py [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        # that may have 'address-map-secure' properties.
+
+        first_cpu_compat = set(_compatible(self.cpus[0]))
+        if not all(set(_compatible(cpu)) == first_cpu_compat
+                   for cpu in self.cpus):
+            _LOG.warning(f"domain '{self.node.path}' CPUs have differing "
+                         'compatible properties; unexpected address map '
+                         'behavior may result')
+
+        def arm_v8m_address_map():
```
Comment (nordicjm) [inline scripts/dts/python-devicetree/src/devicetree/sysdtlib.py:416]: Do we envisage similar requirements for other architectures that would need special handling here? I don't know if arc/mips/ppc/risc/xtensa have or would implement things similar to armv8 security. And also, how would this then support being used/extended out of tree?

File: cmake/modules/dts.cmake (hunk 1 above)
Comment (mbolivar-nordic) [reply cmake/modules/dts.cmake:255]: So far we are trying to be careful to use different extensions for system DT and standard DT files:

- `.sysdts` vs `.dts`
- `.sysdtsi` vs `.dtsi`

So I guess this would be `.sysoverlay`?

File: scripts/dts/python-devicetree/src/devicetree/sysdtlib.py (hunk 3 above)
Comment (mbolivar-nordic) [reply scripts/dts/python-devicetree/src/devicetree/sysdtlib.py:416]: > Do we envisage similar requirements for other architectures that would need special handling here?

I don't know. Anecdotally, this is the first time that the system DT maintainers ran into this; they had no idea this was a possiblility, and they're running on multiple architectures.

I think the answer is probably "yes", but I also think we should proceed incrementally and gather more data with case-by-case code like this until a pattern emerges. It's not like new architectures are appearing every day.

> And also, how would this then support being used/extended out of tree?

Right now, the system devicetree specification itself is where the meaning of the execution level flags is defined, and the spec authors chose to mandate that the spec itself should be updated to add new architecture-specific definitions for this field. (I do plan on upstreaming this Arm v8-M convention into the spec as a replacement for a couple of previous, failed attempts to handle this case that we reverted in the spec; see https://github.com/devicetree-org/lopper/commit/0da7cbd70bc2c3417e5e909bdd90eea2a0a0c912 .)

TL;DR "you won't extend this out of tree for now". Let's gather data from publicly documented architectures that want to support this, then when the pattern emerges, we can extend the specification as needed.


So for now, this is a "heavyweight" operation

**Comment by mbolivar-nordic** [comment]
Last force push is just a rebase

**Comment by mbolivar-nordic** [comment]
I separated the prep work commits out of this PR and into their own pull request:

https://github.com/zephyrproject-rtos/zephyr/pull/55105

In this PR's latest force-push, I just squashed them together, with a commit message saying they can be ignored as they're separately reviewable enablement work.

**Comment by mbolivar-nordic** [comment]
> @tejlmand seems this is not compatible with sysbuild

@nordicjm fixed in latest force push.

I also added sysbuild-aware support for a SYSDT_OVERLAY_FILE which functions analogously to DTC_OVERLAY_FILE for y'all to play with.

**Comment by mbolivar-nordic** [comment]
> I separated the prep work commits out of this PR and into their own pull requests:

These are now merged, so I've rebased their contents out of this PR to make it smaller.

**Comment by mbolivar-nordic** [comment]
This now separates the system DT portions into their own CMake module. The reviewable refactoring that enabled this is in a separate PR: https://github.com/zephyrproject-rtos/zephyr/pull/55293. This RFC is rebased on top of that PR.
